File: widgety/R/widgety.R
#' <Add Title>
#'
#' <Add Description>
#'
#' @import htmlwidgets
#'
#' @export
widgety <- function(message="Hello, world!",
                    radius=100,
                    circle_stroke="black",
                    circle_stroke_width=1,
                    circle_fill="steelblue",
                    family="Lato",
                    text_color="white",
                    width = NULL, height = NULL) {

  # forward options using x
  widget_options = list(
    message = message,
    radius = radius,
    circle_stroke = circle_stroke,
    circle_stroke_width = circle_stroke_width,
    circle_fill = circle_fill,
    family = family,
    text_color = text_color
  )

  # create widget
  htmlwidgets::createWidget(
    name = 'widgety',
    x = widget_options,
    width = width,
    height = height,
    package = 'widgety'
  )
}

#' Widget output function for use in Shiny
#'
#' @export
widgetyOutput <- function(outputId, width = '100%', height = '400px'){
  shinyWidgetOutput(outputId, 'widgety', width, height, package = 'widgety')
}

#' Widget render function for use in Shiny
#'
#' @export
renderWidgety <- function(expr, env = parent.frame(), quoted = FALSE) {
  if (!quoted) { expr <- substitute(expr) } # force quoted
  shinyRenderWidget(expr, widgetyOutput, env, quoted = TRUE)
}

File: widgety/inst/htmlwidgets/widgety.js
HTMLWidgets.widget({

  name: 'widgety',

  type: 'output',

  initialize: function(el, width, height) {

    return {
      w:width,
      h:height
    };

  },

  renderValue: function(el, x, instance) {

    var visDiv = d3.select(el);

    var svg = visDiv.append("svg")
          .attr("width", instance.w)
          .attr("height", instance.h);

    var circle = svg.append("circle")
          .attr("cx", instance.w/2)
          .attr("cy", instance.h/2)
          .attr("r", x.radius)
          .attr("stroke-width", x.circle_stroke_width)
          .attr("stroke", x.circle_stroke)
          .attr("fill", x.circle_fill);

    var label = svg.append("text")
         .attr("x", instance.w/2)
         .attr("y", instance.h/2)
         .attr("text-anchor", "middle")
         .attr("font-family", x.family)
         .attr("fill", x.text_color)
         .text(x.message);

  },

  resize: function(el, width, height, instance) {

  }

});
